0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,180 @@
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
+This allows us to add information to one specific swap without having to update the `swap` table.
+Note that the actual secret is not stored in the database, Alice always computes it from the swap parameters and her seed.
+
+Third we dealt with the problem that there is not always an `address_hint` (i.e. Bob does not need to know the `address_hint` of Alice).
+We added the `address_hint` extension table that relates to `swap` on the `counter_party_peer_id` field (rather than on the `swap_local_id`).
+
+Note that the role could be factored out of the table, but we decided against it because it adds unnecessary complexity for now.
+
+What remains is a simple table for `swap` which contains only the information which all swaps have in common: `role`, `alpha_protocol`, `beta_protocol` and `counter_party_peer_id`.
+
+=== Protocols
+
+[[HanProtocols]]
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/han_swap_protocols.puml&fmt=svg[HanProtocols]
+
+In order to store details for each swap such as data we need for creating the HTLCs for each protocol we introduce `\{protocol\}_swap_details`, where `\{protocol\}` can be `han_bitcoin`, ``han_ethereum`, `herc20` or `halight` for now.
+Similar to the `Swaps` table, we add information which is only known at a later point of time into a sub-table.
+This resulted in quite an excess of sub-tables such as `han_bitcoin/ethereum_redeem_identity`, `han_bitcoin/ethereum_refund_identity`, etc.
+
+Some notable sub-tables:
+
+* `han_bitcoin/ethereum_swap_detail` contains data needed to create the HTLCs. We decided to give this table a generated primary key, rather than using `swap_local_id` as the primary key to allow same chain swaps.
+* `han_bitcoin_ethereum_redeem/refund_identity`: Since some identities are exchanged during announcement, the identities are kept in extension tables.
+
+One could optimize here and combine all ledger specific identities into an own table `han_bitcoin_identity` to hold all Bitcoin `refund` and `redeem` identities.
+This would mean though that we would need an additional unique id in those sub-tables, i.e. we can't use the `fk` as `pk`.
+
+* `han_bitcoin/ethereum_ledger_state`: is a new table which we introduced in this database schema.
+It records observed ledger events such as `fund`/`redeem`/`refund`-transaction.
+We extracted the states into `han_state` in order to reduce duplication in each table and not have an extra column for the event-type (`fund`/`redeem`/`refund`).
```

Comment:
Given that each even may come with different data, my initial guess would be that we should have one table per event, with a FK relation to the protocol (i.e. a row in the `han_bitcoin_protocol` table).

Reply:
Oh right. we did not consider the additional data for the states but only considered the tx_ids.

Reply:
We generally don't model the event properly, but mesh the event into the state and just track the tx ids. I wonder how far we want to take this. 
The idea is that the blockchain remains the base of truth, but the tx_id helps us to speed things up. (@bonomat correct me if I'm wrong :)

Reply:
events != state.

There is always just one state but multiple events.
If you want to model states, you need to think in `UPDATE` statements.
I think modelling events makes more sense here as they are the source of truth.

In any case, all of this is out of scope anyway. I should have never commented on it in the first place because it doesn't matter at the moment :sweat_smile:

Reply:
I mark this as resolved. Please reopen and comment if you disagree.